clickhouse-admin/types/src/lib.rs

```diff
@@ -206,21 +209,198 @@ impl KeeperSettings {
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use std::{
-        net::{Ipv4Addr, Ipv6Addr},
-        str::FromStr,
+macro_rules! define_struct_and_set_values {
+    (
+        #[$doc_struct:meta]
+        struct $name:ident {
+            $(
+                #[$doc_field:meta]
+                $field_name:ident: $field_type:ty
+            ),* $(,)?
+        }
+    ) => {
+        #[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, JsonSchema)]
+        #[serde(rename_all = "snake_case")]
+        #[$doc_struct]
+        pub struct $name {
+            $(
+                #[$doc_field]
+                $field_name: $field_type
+            ),*
+        }
+
+        impl $name {
+            // Check if a field name matches a given key and set its value
+            pub fn set_field_value(&mut self, key: &str, value: Option<u64>) -> Result<()> {
+                match key {
+                    $(
+                        stringify!($field_name) => {
+                            self.$field_name = value;
+                            Ok(())
+                        },
+                    )*
+                    _ => bail!("Field name '{}' not found.", key),
+                }
+            }
+
+            fn are_all_fields_none(&self) -> bool {
+                $(
+                    self.$field_name.is_none() &&
+                )*
+                true
+            }
+        }
     };
+}
 
+define_struct_and_set_values! {
+    /// Logically grouped information file from a keeper node
+    struct Lgif {
+        /// Index of the first log entry in the current log segment
+        first_log_idx: Option<u64>,
+        /// Term of the leader when the first log entry was created
+        first_log_term: Option<u64>,
+        /// Index of the last log entry in the current log segment
+        last_log_idx: Option<u64>,
+        /// Term of the leader when the last log entry was created
+        last_log_term: Option<u64>,
+        /// Index of the last committed log entry
+        last_committed_log_idx: Option<u64>,
+        /// Index of the last committed log entry from the leader's perspective
+        leader_committed_log_idx: Option<u64>,
+        /// Target index for log commitment during replication or recovery
+        target_committed_log_idx: Option<u64>,
+        /// Index of the most recent snapshot taken
+        last_snapshot_idx: Option<u64>,
+    }
+}
+
+impl Lgif {
+    pub fn new() -> Self {
+        Self {
+            first_log_idx: None,
+            first_log_term: None,
+            last_log_idx: None,
+            last_log_term: None,
+            last_committed_log_idx: None,
+            leader_committed_log_idx: None,
+            target_committed_log_idx: None,
+            last_snapshot_idx: None,
+        }
+    }
+
+    pub fn parse(log: &Logger, data: &[u8]) -> Result<Self> {
```

Comment:
Hm, ok, let me play around with this a bit and simplify it :D

Reply:
Thank you :)

Reply:
I changed the code to return errors instead of trying to fill in as much of the response as it could.

I tried to get rid of the macro but hear me out 😄 . The moment I pass the raw output through `.lines()` the order of the items gets completely changed, which means order expected parsing could be error prone. On the other hand, the macro gives us assurance that a field of the struct will only be populated if the struct filed name is _exactly_ the same as the key from the command output. I can't really think of a safer way to do this, but open to suggestions!

Reply:
> The moment I pass the raw output through .lines() the order of the items gets completely changed

This is not how `lines` works. It's a lazy iterator like all others and returns one line at a time in the current order. Are you sure the change in order isn't because you are inserting into a `HashMap` to collect the lines? There's also a sort below that will change the order. To keep the same order you probably want to collect into a `Vec` instead.